posts/2018-11-28-http4s-error-handling-mtl-2.md

```diff
@@ -0,0 +1,456 @@
+    users: UserAlg[F, E]
+) extends Routes[F, E]
+```
+
+That's it! We are now enforcing this relationship at compile time. Let's see how the `HttpRoutes` looks like:
+
+```tut:book:silent
+class UserRoutesAlt[F[_]: HttpErrorHandler[?[_], UserError]: Sync](
+    users: UserAlg[F, UserError]
+) extends UserRoutes(users) {
+
+  protected val httpRoutes: HttpRoutes[F] = HttpRoutes.of[F] {
+
+    case GET -> Root / "users" / username =>
+      users.find(username).flatMap {
+        case Some(user) => Ok(user.asJson)
+        case None       => NotFound(username.asJson)
+      }
+
+    case req @ POST -> Root / "users" =>
+      req
+        .as[User]
+        .flatMap { user =>
+          users.save(user) *> Created(user.username.asJson)
+        }
+
+    case req @ PUT -> Root / "users" / username =>
+      req
+        .as[UserUpdateAge]
+        .flatMap { userUpdate =>
+          users.updateAge(username, userUpdate.age) *> Ok(username.asJson)
+        }
+  }
+
+}
+```
+
+Neat! Right? If we try to change the error type of `UserAlg` it wouldn't compile!
+
+### More than one algebra per Http Route
+
+In most of my programs I tend to specify an `HttpRoute` per algebra. But what if we wanted to just define a single `HttpRoute` that uses multiple algebras? There are a couple of options.
+
+Let's first define a new ADT of errors and a new algebra to illustrate the problem:
+
+**Catalog Error**
+
+```tut:book:silent
+sealed trait CatalogError extends Exception
+case class ItemAlreadyExists(item: String) extends CatalogError
+case class CatalogNotFound(id: Long) extends CatalogError
+```
+
+**CatalogAlg**
+
+```tut:book:silent
+case class Item(name: String) extends AnyVal
+
+abstract class CatalogAlg[F[_]: ErrorChannel[?[_], E], E <: Throwable] {
+  def find(id: Long): F[List[Item]]
+  def save(id: Long, item: Item): F[Unit]
+}
+```
+
+**HttpRoutes with multiple algebras**
+
+Here we have an `HttpRoutes` that makes use of two algebras with different error types:
+
+```tut:book:silent
+class UserRoutesMTL[F[_]: Sync](
+    users: UserAlg[F, UserError],
+    catalog: CatalogAlg[F, CatalogError]
+) extends Http4sDsl[F] {
+
+  private val httpRoutes: HttpRoutes[F] = ???
+
+  def routes(
+    implicit H1: HttpErrorHandler[F, UserError],
+             H2: HttpErrorHandler[F, CatalogError]
+  ): HttpRoutes[F] =
+    H2.handle(H1.handle(httpRoutes))
+
+}
+```
+
+It works! But it's not as elegant as we would like it to be and if we add more algebras this would quicky get out of control.
+
+***Can we generalize this pattern?***
+
+### Shapeless Coproduct
+
+We can define our error type as a coproduct of different errors, in our case `UserError` and `CatalogError`. For example:
+
+```scala
+import shapeless._
+
+def routes[F[_]](implicit H: HttpErrorHandler[F, UserError :+: CatalogError :+: CNil]) = ???
+```
+
+However, this doesn't compile because the error type is no longer a subtype of `Throwable`. It is now a `Coproduct` :/. But...
```

Comment:
Neither ASCII emojis nor ellipses are particularly good style. Try to rephrase it slightly, e.g. join with the next paragraph.